core/src/test/java/de/danoeh/antennapod/core/storage/ItemEnqueuePositionCalculatorTest.java

```diff
@@ -0,0 +1,313 @@
+
+            FeedItem tFI201 = tFI_isDownloading(201, stubDownloadStateProvider);
+            doAddToQueueAndAssertResult(message + " (bulk insertion, 1st item)",
+                    calculator, 0, tFI201, queue,
+                    idsExpectedAfter201);
+
+            FeedItem tFI202 = tFI_isDownloading(202, stubDownloadStateProvider);
+            doAddToQueueAndAssertResult(message + " (bulk insertion, 2nd item)",
+                    calculator, 1, tFI202, queue,
+                    idsExpectedAfter202);
+
+            // TODO: simulate download failure cases.
+        }
+
+
+        private static FeedItem tFI_isDownloading(int id, DownloadStateProvider stubDownloadStateProvider) {
+            FeedItem item = tFI(id);
+            FeedMedia media =
+                    new FeedMedia(item, "http://download.url.net/" + id
+                            , 100000 + id, "audio/mp3");
+            media.setId(item.getId());
+            item.setMedia(media);
+
+            stub(stubDownloadStateProvider.isDownloadingFile(media)).toReturn(true);
+
+            return item;
+        }
+    }
+
+
+    // Common helpers:
+    // - common queue (of items) for tests
+    // - construct FeedItems for tests
+
+    static void doAddToQueueAndAssertResult(String message,
+                                            ItemEnqueuePositionCalculator calculator,
+                                            int positionAmongAdd,
+                                            FeedItem itemToAdd,
+                                            List<FeedItem> queue,
+                                            List<Long> idsExpected) {
+        int posActual = calculator.calcPosition(positionAmongAdd, itemToAdd, queue);
+        queue.add(posActual, itemToAdd);
+        assertEquals(message, idsExpected, toIDs(queue));
+    }
+
+    static final List<FeedItem> QUEUE_EMPTY = Collections.unmodifiableList(Arrays.asList());
+
+    static final List<FeedItem> QUEUE_DEFAULT = Collections.unmodifiableList(Arrays.asList(tFI(11), tFI(12), tFI(13), tFI(14)));
+    static final List<Long> QUEUE_DEFAULT_IDS = QUEUE_DEFAULT.stream().map(fi -> fi.getId()).collect(Collectors.toList());
+
+
+    static FeedItem tFI(long id) {
+        return tFI(id, -1);
+    }
+
+    static FeedItem tFI(long id, int position) {
+        FeedItem item = tFINoMedia(id);
+        FeedMedia media = new FeedMedia(item, "download_url", 1234567, "audio/mpeg");
+        media.setId(item.getId());
+        item.setMedia(media);
+
+        if (position >= 0) {
+            media.setPosition(position);
+        }
+
+        return item;
+    }
+
+    static FeedItem tFINoMedia(long id) {
+        FeedItem item = new FeedItem(id, "Item" + id, "ItemId" + id, "url",
+                new Date(), FeedItem.PLAYED, FeedMother.anyFeed());
+        return item;
+    }
+
+    // Collections helpers
+
+    static <T> List<? extends T> concat(T item, List<? extends T> list) {
```

Comment:
Those look like they should be in another class. They are not specific to `ItemEnqueuePositionCalculatorTest`

Reply:
What about a new `core.util.CollectionUtil`? Should it stay in unit test or be part of the main code base?

Reply:
I don't think we currently need them in the main code base.